src/jsonrpc.rs

```diff
@@ -156,26 +190,69 @@ fn process_http_request(obj: JsonRequest, infura_url: &str) -> Vec<u8> {
     }
 }
 
-fn handle_request(obj: JsonRequest, infura_url: &str) -> Result<String, String> {
-    let request_id = obj.id;
-    match obj.method.as_str() {
-        "web3_clientVersion" => Ok(format!(
-            r#"{{"jsonrpc":"2.0","id":{},"result":"trin 0.0.1-alpha"}}"#,
-            request_id,
+fn handle_request(
+    obj: JsonRequest,
+    infura_url: &str,
+    portal_tx: Sender<PortalEndpoint>,
+) -> Result<String, String> {
+    let method = obj.method.as_str();
+    if method.starts_with("discv5") {
+        dispatch_portal_request(obj, portal_tx)
+    } else {
+        match method {
+            "web3_clientVersion" => Ok(format!(
+                r#"{{"jsonrpc":"2.0","id":{},"result":"trin 0.0.1-alpha"}}"#,
+                obj.id,
+            )),
+            _ => dispatch_infura_request(obj, infura_url),
+        }
+    }
+}
+
+fn dispatch_infura_request(obj: JsonRequest, infura_url: &str) -> Result<String, String> {
+    //Re-encode json to proxy to Infura
+    let request = serde_json::to_string(&obj).unwrap();
+    match proxy_to_url(request, infura_url) {
+        Ok(result_body) => Ok(std::str::from_utf8(&result_body).unwrap().to_owned()),
+        Err(err) => Err(format!(
+            r#"{{"jsonrpc":"2.0","id":"{}","error":"Infura failure: {}"}}"#,
+            obj.id,
+            err.to_string(),
         )),
+    }
+}
+
+#[tokio::main]
+async fn dispatch_portal_request(
+    obj: JsonRequest,
+    portal_tx: Sender<PortalEndpoint>,
+) -> Result<String, String> {
+    let (resp_tx, resp_rx) = oneshot::channel::<Result<Value, String>>();
+    let method = obj.method.as_str();
+    let message = match method {
+        "discv5_nodeInfo" => PortalEndpoint::NodeInfo { resp: resp_tx },
+        "discv5_routingTableInfo" => PortalEndpoint::RoutingTableInfo { resp: resp_tx },
         _ => {
-            //Re-encode json to proxy to Infura
-            let request = serde_json::to_string(&obj).unwrap();
-            match proxy_to_url(request, infura_url) {
-                Ok(result_body) => Ok(std::str::from_utf8(&result_body).unwrap().to_owned()),
-                Err(err) => Err(format!(
-                    r#"{{"jsonrpc":"2.0","id":"{}","error":"Infura failure: {}"}}"#,
-                    request_id,
-                    err.to_string(),
-                )),
-            }
+            return Err(format!(
+                r#"{{"jsonrpc":"2.0","id":"{}","error":"Unsupported discv5 endpoint: {}"}}"#,
+                obj.id, method,
+            ))
         }
-    }
+    };
+    tokio::task::spawn(async move { portal_tx.send(message).await });
```

Comment:
Alternatively, you could use an [`UnboundedSender`](https://docs.rs/tokio/1.8.1/tokio/sync/mpsc/struct.UnboundedSender.html), which doesn't require an `await` to send.